Objective-C/Tests/ReplicatorTest.m

```diff
@@ -1855,6 +1855,146 @@ - (void) testPushAndForget {
+    
+    // Check documents passed to the filter:
+    AssertEqual(docIds.count, 2u);
+    Assert([docIds containsObject: @"doc1"]);
+    Assert([docIds containsObject: @"pass"]);
+    
+    // Check replicated documents:
+    AssertNotNil([otherDB documentWithID: @"doc1"]);
+    AssertNotNil([otherDB documentWithID: @"pass"]);
+    
+    Assert([self.db deleteDocument: doc1 error: &error]);
+    Assert([self.db deleteDocument: doc2 error: &error]);
+    
+    [docIds removeAllObjects];
+    [self run: config errorCode: 0 errorDomain: nil];
+    
+    AssertEqual(docIds.count, 2u);
+    Assert([docIds containsObject: @"doc1"]);
+    Assert([docIds containsObject: @"pass"]);
+    
+    // shouldn't delete the one with `docID != pass`
+    AssertNotNil([otherDB documentWithID: @"doc1"]);
+    AssertNil([otherDB documentWithID: @"pass"]);
+}
+
+
+- (void) testPullDeletedDocWithFilter: (BOOL)isContinuous {
+    // Create documents:
+    NSError* error;
+    CBLMutableDocument* doc1 = [[CBLMutableDocument alloc] initWithID: @"doc1"];
+    [doc1 setString: @"pass" forKey: @"name"];
+    Assert([otherDB saveDocument: doc1 error: &error]);
+    
+    CBLMutableDocument* pass = [[CBLMutableDocument alloc] initWithID: @"pass"];
+    [pass setString: @"pass" forKey: @"name"];
+    Assert([otherDB saveDocument: pass error: &error]);
+    
+    NSMutableSet<NSString*>* docIds = [NSMutableSet set];
+    id target = [[CBLDatabaseEndpoint alloc] initWithDatabase: otherDB];
+    CBLReplicatorConfiguration* config = [self configWithTarget: target
+                                                           type: kCBLReplicatorTypePull
+                                                     continuous: isContinuous];
+    config.pullFilter = ^BOOL(CBLDocument* document, CBLDocumentFlags flags) {
+        AssertNotNil(document.id);
+        [docIds addObject: document.id];
+        
+        BOOL isDeleted = (flags & kCBLDocumentFlagsDeleted) == kCBLDocumentFlagsDeleted;
+        if (isDeleted) {
+            // if deleted only allow  `docID = pass` is allowed.
+            return [document.id isEqualToString: @"pass"];
+        }
+        // allow all docs with `name = pass`
+        return [[document stringForKey: @"name"] isEqualToString: @"pass"];
+    };
+    
+    [self run: config errorCode: 0 errorDomain: nil];
+    
+    // Check documents passed to the filter:
+    AssertEqual(docIds.count, 2u);
+    Assert([docIds containsObject: @"doc1"]);
+    Assert([docIds containsObject: @"pass"]);
+    
+    // should replicate all docs with `name = pass`
+    AssertNotNil([self.db documentWithID: @"doc1"]);
+    AssertNotNil([self.db documentWithID: @"pass"]);
+    AssertEqual(self.db.count, 2u);
+    AssertEqual(otherDB.count, 2u);
+    
+    Assert([otherDB deleteDocument: doc1 error: &error]);
+    Assert([otherDB deleteDocument: pass error: &error]);
+    
+    [self run: config errorCode: 0 errorDomain: nil];
```

Comment:
Better to reset the docIds before re-running the replicator.

Reply:
now only checking for deleted docIds, since the other checking is already doing in other tests. We are validating whether the deleted flag is not added on the first run.

Reply:
done!